src/GridFieldActionsMenu.php

```diff
@@ -0,0 +1,277 @@
+     */
+    protected $showFirstTab = true;
+
+    /**
+     * The action items to show in the menu
+     *
+     * Example structure:
+     *
+     * <code>
+     * // Actions
+     * [
+     *     // Group
+     *     [
+     *         // Action
+     *         [
+     *             'Title' => 'Content',
+     *             'Link' => 'hrefme',
+     *             'Type' => 'arbitrary_name', // e.g. 'link' or 'versioning'
+     *         ],
+     *         ...
+     *     ],
+     *     // or a callable that generates the above Group level array format
+     *     function($gridField, $record, $typeName) ...
+     *     ...
+     * ]
+     * </code>
+     *
+     * @var array
+     */
+    protected $actions = [];
+
+    public function __construct($showFirstTab = true)
+    {
+        $this->setShowFirstTab($showFirstTab);
+        $this->actions = [
+            'link' => [$this, 'addRootTabActions'],
+            'versioning' => [$this, 'addVersionedActions']
+        ];
+    }
+
+    public function augmentColumns($gridField, &$columns)
+    {
+        if (!in_array('Actions', $columns)) {
+            $columns[] = 'Actions';
+        }
+    }
+
+    public function getColumnsHandled($gridField)
+    {
+        return ['Actions'];
+    }
+
+    /**
+     * Add each of the "Root" tabs to the actions for this component
+     * We expect that a tabbed list of fields will always have a singular root.
+     *
+     * @param DataObject $record
+     * @param string $typeName
+     *
+     * @return array
+     */
+    protected function addRootTabActions(GridField $gridField, DataObject $record, $typeName)
+    {
+        $tabSet = $record->getCMSFields()->first();
+        if (!($tabSet instanceof TabSet)) {
+            return [];
+        }
+
+        $first = true;
+        $actions = [];
+        foreach ($tabSet->Tabs() as $tab) {
+            /** @var \SilverStripe\Forms\Tab $tab */
+            $tabID = ($first) ? null : $tab->ID();
+            $link = Controller::join_links($gridField->Link('item'), $record->ID, 'edit');
+            // @TODO hack workaround: `&& false` here because some JS in the CMS is rewriting
+            // a link with a hash in it to the page we're on _now_ #anchor, as opposed to
+            // e.g link/set/here#anchor
+            $link = $tabID && false ? "$link#$tabID" : $link;
+
+            $actions[] = [
+                'Title' => $tab->Title(),
+                'Link' => $link,
+                'Type' => $typeName
+            ];
+            $first = false;
+        }
+        return $actions;
+    }
+
+    /**
+     * If the object is versioned (has the {@link Versioned} extension applied) then add
+     * actions to publish, unpublish, and archive - depending on record state.
+     *
+     * @param DataObject $record
+     * @param string $typeName
+     *
+     * @return array
+     */
+    protected function addVersionedActions(GridField $gridField, DataObject $record, $typeName)
```

Comment:
Doesn't match phpdoc

Reply:
Fixed. Comment on line that hasn't changed, so doesn't "auto-resolve".